app/ide-desktop/lib/client/src/url-associations.ts

```diff
@@ -0,0 +1,180 @@
+ * is called.
+ *
+ * @param callback - The callback to call when the application is requested to open a URL.
+ */
+export function registerUrlCallback(callback: (url: URL) => void) {
+    // First, register the callback for the `open-url` event. This is used on macOS.
+    electron.app.on('open-url', (event, url) => {
+        logger.log(`Got URL from 'open-url' event: '${url}'.`)
+        event.preventDefault()
+        callback(new URL(url))
+    })
+
+    // Second, register the callback for the `second-instance` event. This is used on Windows.
+    electron.app.on('second-instance', (event, argv) => {
+        logger.log(`Got data from 'second-instance' event: '${argv.toString()}'.`)
+        unsetAsUrlHandler()
+        // Check if additional data is an object that contains the URL.
+        const requestOneLastElementSlice = -1
+        const lastArgumentSlice = argv.slice(requestOneLastElementSlice)
+        const url = argsDenoteUrlOpenAttempt(lastArgumentSlice)
+        if (url) {
+            logger.log(`Got URL from 'second-instance' event: '${url.toString()}'.`)
+            // Even we received the URL, our Window likely is not in the foreground - the focus
+            // went to the "second instance" of the application. We must bring our Window to the
+            // foreground, so the user gets back to the IDE after the authentication.
+            const primaryWindow = electron.BrowserWindow.getAllWindows()[0]
+            if (primaryWindow) {
+                if (primaryWindow.isMinimized()) {
+                    primaryWindow.restore()
+                }
+                primaryWindow.focus()
+            } else {
+                logger.error('No primary window found after receiving URL from second instance.')
+            }
+            logger.log(`Got URL from second instance: '${url.toString()}'.`)
+            event.preventDefault()
+            callback(url)
+        }
+    })
+}
+
+// ===============================
+// === Temporary handler setup ===
+// ===============================
+
+/** Make this application instance the recipient of URL callbacks.
+ *
+ * After the callback is received (or no longer expected), the `urlCallbackCompleted` function
+ * must be called. Otherwise, other IDE instances will not be able to receive their URL
+ * callbacks.
+ *
+ * The mechanism is built on top of the Electron's
+ * [instance lock]{@link https://www.electronjs.org/docs/api/app#apprequestsingleinstancelock} functionality.
+ *
+ * @throws An error if another instance of the application has already acquired the lock.
+ */
+export function setAsUrlHandler() {
+    logger.log('Expecting URL callback, acquiring the lock.')
+    if (!electron.app.requestSingleInstanceLock()) {
+        const message = 'Another instance of the application is already running. Exiting.'
+        logger.error(message)
+        // eslint-disable-next-line no-restricted-syntax
```

Comment:
this should be redundant too

Reply:
This one seems to be still required.

Reply:
ah, yeah, my bad - i think you can suppress it by explicitly doing `else { return }` though